src/betterproto/__init__.py

```diff
@@ -505,30 +545,54 @@ class Message(ABC):
     _serialized_on_wire: bool
     _unknown_fields: bytes
     _group_current: Dict[str, str]
+    __dataclass_fields__: Mapping[str, dataclasses.Field]  # Technically should be Final
+    # as this is not settable by the end user, but that doesn't come till python 3.8
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        args = list(args)
+        set_attribute = super().__setattr__  # Save a super() call every time
+        all_sentinel = True  # Keep track of whether every field was default
+        group_current: Dict[str, str] = {}
 
-    def __post_init__(self) -> None:
-        # Keep track of whether every field was default
-        all_sentinel = True
+        for field in self.__raw_get(dataclasses._FIELDS).values():
+            # We have to set these here to allow for them to be editable
+            set_attribute(field.name, field.default)
 
-        # Set current field of each group after `__init__` has already been run.
-        group_current: Dict[str, Optional[str]] = {}
         for field_name, meta in self._betterproto.meta_by_field_name.items():
+            if args or kwargs:
+                try:
+                    set_attribute(
+                        field_name, (args or kwargs).pop(0 if args else field_name)
```

Comment:
this is a bit cryptic. Maybe having more distinct code paths for args vs kwargs would be better. In fact I think the control flow within this loop is worth revisiting more generally.

Reply:
This was the fastest way I could think of to make the `__init__` run, if you aren't a fan of this I'll do some exec grossness to do in effect what happens in the code block in https://docs.python.org/3/library/dataclasses.html

Reply:
I think it would be better rather than doing what I'm doing ATM to make generated protos have a generated `__init__`?

Reply:
As currently the code is really really gross 
```py
    def __new__(
        mcs, name: str, bases: Tuple[type, ...], attrs: Dict[str, Any]
    ) -> Type["Message"]:
        annotations = attrs.get("__annotations__", {})
        attrs["__slots__"] = tuple(annotations)
        # Slot the class

        fields = {}
        for annotation, type in annotations.items():
            field: Optional[dataclasses.Field] = attrs.pop(annotation, None)
            # Remove field class variables from the class namespace
            if field is not None:
                field.name = annotation
                field.type = type
                field._field_type = dataclasses._FIELD
                fields[annotation] = field

        message_class = super().__new__(mcs, name, bases, attrs)
        message_class.__dataclass_fields__ = fields
        body = "\n".join(
            f"    self.{field_name} = {field_name}" for field_name in fields
        )
        exec(
            f"""
def __init__(self, {", ".join(f"{field_name} = PLACEHOLDER" for field_name in fields)}):
    object.__setattr__(self, "_group_current", {'{}'})
    object.__setattr__(self, "_init_running", True)
{body}
    # Now that all the defaults are set, reset it!
    object.__setattr__(self, "_unknown_fields", b"")
    object.__setattr__(self, "_init_running", False)
    object.__setattr__(self, "_serialized_on_wire", False)
message_class.__init__ = __init__
"""
        )
        return message_class
```

Reply:
Although having the init be per class kind of defeats the purpose of originally having these as dataclasses so maybe this is the only way to do it

Reply:
I'm not sure what the best solution is just by looking at it, there are clearly a few reasonable approaches. My issue with the existing implementation is just that the control flow could be cleaner.